// agentmail/dist/esm/Client.mjs
// This file was auto-generated by Fern from our API Definition.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { AgentClient } from "./api/resources/agent/client/Client.mjs";
import { ApiKeysClient } from "./api/resources/apiKeys/client/Client.mjs";
import { AuthClient } from "./api/resources/auth/client/Client.mjs";
import { DomainsClient } from "./api/resources/domains/client/Client.mjs";
import { DraftsClient } from "./api/resources/drafts/client/Client.mjs";
import { InboxesClient } from "./api/resources/inboxes/client/Client.mjs";
import { ListsClient } from "./api/resources/lists/client/Client.mjs";
import { MetricsClient } from "./api/resources/metrics/client/Client.mjs";
import { OrganizationsClient } from "./api/resources/organizations/client/Client.mjs";
import { PodsClient } from "./api/resources/pods/client/Client.mjs";
import { ThreadsClient } from "./api/resources/threads/client/Client.mjs";
import { WebhooksClient } from "./api/resources/webhooks/client/Client.mjs";
import { WebsocketsClient } from "./api/resources/websockets/client/Client.mjs";
import { normalizeClientOptionsWithAuth } from "./BaseClient.mjs";
import * as core from "./core/index.mjs";
import * as environments from "./environments.mjs";
export class AgentMailClient {
    constructor(options = {}) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    get inboxes() {
        var _a;
        return ((_a = this._inboxes) !== null && _a !== void 0 ? _a : (this._inboxes = new InboxesClient(this._options)));
    }
    get pods() {
        var _a;
        return ((_a = this._pods) !== null && _a !== void 0 ? _a : (this._pods = new PodsClient(this._options)));
    }
    get webhooks() {
        var _a;
        return ((_a = this._webhooks) !== null && _a !== void 0 ? _a : (this._webhooks = new WebhooksClient(this._options)));
    }
    get agent() {
        var _a;
        return ((_a = this._agent) !== null && _a !== void 0 ? _a : (this._agent = new AgentClient(this._options)));
    }
    get apiKeys() {
        var _a;
        return ((_a = this._apiKeys) !== null && _a !== void 0 ? _a : (this._apiKeys = new ApiKeysClient(this._options)));
    }
    get auth() {
        var _a;
        return ((_a = this._auth) !== null && _a !== void 0 ? _a : (this._auth = new AuthClient(this._options)));
    }
    get domains() {
        var _a;
        return ((_a = this._domains) !== null && _a !== void 0 ? _a : (this._domains = new DomainsClient(this._options)));
    }
    get drafts() {
        var _a;
        return ((_a = this._drafts) !== null && _a !== void 0 ? _a : (this._drafts = new DraftsClient(this._options)));
    }
    get lists() {
        var _a;
        return ((_a = this._lists) !== null && _a !== void 0 ? _a : (this._lists = new ListsClient(this._options)));
    }
    get metrics() {
        var _a;
        return ((_a = this._metrics) !== null && _a !== void 0 ? _a : (this._metrics = new MetricsClient(this._options)));
    }
    get organizations() {
        var _a;
        return ((_a = this._organizations) !== null && _a !== void 0 ? _a : (this._organizations = new OrganizationsClient(this._options)));
    }
    get threads() {
        var _a;
        return ((_a = this._threads) !== null && _a !== void 0 ? _a : (this._threads = new ThreadsClient(this._options)));
    }
    get websockets() {
        var _a;
        return ((_a = this._websockets) !== null && _a !== void 0 ? _a : (this._websockets = new WebsocketsClient(this._options)));
    }
    /**
     * Make a passthrough request using the SDK's configured auth, retry, logging, etc.
     * This is useful for making requests to endpoints not yet supported in the SDK.
     * The input can be a URL string, URL object, or Request object. Relative paths are resolved against the configured base URL.
     *
     * @param {Request | string | URL} input - The URL, path, or Request object.
     * @param {RequestInit} init - Standard fetch RequestInit options.
     * @param {core.PassthroughRequest.RequestOptions} requestOptions - Per-request overrides (timeout, retries, headers, abort signal).
     * @returns {Promise<Response>} A standard Response object.
     */
    fetch(input, init, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            return core.makePassthroughRequest(input, init, {
                baseUrl: (_a = this._options.baseUrl) !== null && _a !== void 0 ? _a : (() => __awaiter(this, void 0, void 0, function* () {
                    var _a;
                    const env = yield core.Supplier.get(this._options.environment);
                    return typeof env === "string"
                        ? env
                        : ((_a = env === null || env === void 0 ? void 0 : env.http) !== null && _a !== void 0 ? _a : environments.AgentMailEnvironment.Prod.http);
                })),
                headers: this._options.headers,
                timeoutInSeconds: this._options.timeoutInSeconds,
                maxRetries: this._options.maxRetries,
                fetch: this._options.fetch,
                logging: this._options.logging,
                getAuthHeaders: () => __awaiter(this, void 0, void 0, function* () { return (yield this._options.authProvider.getAuthRequest()).headers; }),
            }, requestOptions);
        });
    }
}
