// agentmail/dist/cjs/serialization/resources/apiKeys/types/ApiKeyPermissions.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiKeyPermissions = void 0;
const core = __importStar(require("../../../../core/index.js"));
exports.ApiKeyPermissions = core.serialization.object({
    inboxRead: core.serialization.property("inbox_read", core.serialization.boolean().optional()),
    inboxCreate: core.serialization.property("inbox_create", core.serialization.boolean().optional()),
    inboxUpdate: core.serialization.property("inbox_update", core.serialization.boolean().optional()),
    inboxDelete: core.serialization.property("inbox_delete", core.serialization.boolean().optional()),
    threadRead: core.serialization.property("thread_read", core.serialization.boolean().optional()),
    threadDelete: core.serialization.property("thread_delete", core.serialization.boolean().optional()),
    messageRead: core.serialization.property("message_read", core.serialization.boolean().optional()),
    messageSend: core.serialization.property("message_send", core.serialization.boolean().optional()),
    messageUpdate: core.serialization.property("message_update", core.serialization.boolean().optional()),
    labelSpamRead: core.serialization.property("label_spam_read", core.serialization.boolean().optional()),
    labelBlockedRead: core.serialization.property("label_blocked_read", core.serialization.boolean().optional()),
    labelTrashRead: core.serialization.property("label_trash_read", core.serialization.boolean().optional()),
    draftRead: core.serialization.property("draft_read", core.serialization.boolean().optional()),
    draftCreate: core.serialization.property("draft_create", core.serialization.boolean().optional()),
    draftUpdate: core.serialization.property("draft_update", core.serialization.boolean().optional()),
    draftDelete: core.serialization.property("draft_delete", core.serialization.boolean().optional()),
    draftSend: core.serialization.property("draft_send", core.serialization.boolean().optional()),
    webhookRead: core.serialization.property("webhook_read", core.serialization.boolean().optional()),
    webhookCreate: core.serialization.property("webhook_create", core.serialization.boolean().optional()),
    webhookUpdate: core.serialization.property("webhook_update", core.serialization.boolean().optional()),
    webhookDelete: core.serialization.property("webhook_delete", core.serialization.boolean().optional()),
    domainRead: core.serialization.property("domain_read", core.serialization.boolean().optional()),
    domainCreate: core.serialization.property("domain_create", core.serialization.boolean().optional()),
    domainUpdate: core.serialization.property("domain_update", core.serialization.boolean().optional()),
    domainDelete: core.serialization.property("domain_delete", core.serialization.boolean().optional()),
    listEntryRead: core.serialization.property("list_entry_read", core.serialization.boolean().optional()),
    listEntryCreate: core.serialization.property("list_entry_create", core.serialization.boolean().optional()),
    listEntryDelete: core.serialization.property("list_entry_delete", core.serialization.boolean().optional()),
    metricsRead: core.serialization.property("metrics_read", core.serialization.boolean().optional()),
    apiKeyRead: core.serialization.property("api_key_read", core.serialization.boolean().optional()),
    apiKeyCreate: core.serialization.property("api_key_create", core.serialization.boolean().optional()),
    apiKeyDelete: core.serialization.property("api_key_delete", core.serialization.boolean().optional()),
    podRead: core.serialization.property("pod_read", core.serialization.boolean().optional()),
    podCreate: core.serialization.property("pod_create", core.serialization.boolean().optional()),
    podDelete: core.serialization.property("pod_delete", core.serialization.boolean().optional()),
});
