// agentmail/dist/esm/serialization/resources/inboxEvents/types/InboxEvent.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../../core/index.mjs";
import { OrganizationId } from "../../../types/OrganizationId.mjs";
import { InboxId } from "../../inboxes/types/InboxId.mjs";
import { InboxEventId } from "./InboxEventId.mjs";
import { InboxEventType } from "./InboxEventType.mjs";
export const InboxEvent = core.serialization.object({
    organizationId: core.serialization.property("organization_id", OrganizationId),
    podId: core.serialization.property("pod_id", core.serialization.string()),
    inboxId: core.serialization.property("inbox_id", InboxId),
    eventId: core.serialization.property("event_id", InboxEventId),
    eventType: core.serialization.property("event_type", InboxEventType),
    messageId: core.serialization.property("message_id", core.serialization.string()),
    label: core.serialization.string(),
    eventAt: core.serialization.property("event_at", core.serialization.date()),
    createdAt: core.serialization.property("created_at", core.serialization.date()),
});
