// agentmail/dist/esm/serialization/resources/apiKeys/types/ApiKeyPermissions.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../../core/index.mjs";
export const ApiKeyPermissions = core.serialization.object({
    inboxRead: core.serialization.property("inbox_read", core.serialization.boolean().optional()),
    inboxCreate: core.serialization.property("inbox_create", core.serialization.boolean().optional()),
    inboxUpdate: core.serialization.property("inbox_update", core.serialization.boolean().optional()),
    inboxDelete: core.serialization.property("inbox_delete", core.serialization.boolean().optional()),
    threadRead: core.serialization.property("thread_read", core.serialization.boolean().optional()),
    threadDelete: core.serialization.property("thread_delete", core.serialization.boolean().optional()),
    messageRead: core.serialization.property("message_read", core.serialization.boolean().optional()),
    messageSend: core.serialization.property("message_send", core.serialization.boolean().optional()),
    messageUpdate: core.serialization.property("message_update", core.serialization.boolean().optional()),
    labelSpamRead: core.serialization.property("label_spam_read", core.serialization.boolean().optional()),
    labelBlockedRead: core.serialization.property("label_blocked_read", core.serialization.boolean().optional()),
    labelTrashRead: core.serialization.property("label_trash_read", core.serialization.boolean().optional()),
    draftRead: core.serialization.property("draft_read", core.serialization.boolean().optional()),
    draftCreate: core.serialization.property("draft_create", core.serialization.boolean().optional()),
    draftUpdate: core.serialization.property("draft_update", core.serialization.boolean().optional()),
    draftDelete: core.serialization.property("draft_delete", core.serialization.boolean().optional()),
    draftSend: core.serialization.property("draft_send", core.serialization.boolean().optional()),
    webhookRead: core.serialization.property("webhook_read", core.serialization.boolean().optional()),
    webhookCreate: core.serialization.property("webhook_create", core.serialization.boolean().optional()),
    webhookUpdate: core.serialization.property("webhook_update", core.serialization.boolean().optional()),
    webhookDelete: core.serialization.property("webhook_delete", core.serialization.boolean().optional()),
    domainRead: core.serialization.property("domain_read", core.serialization.boolean().optional()),
    domainCreate: core.serialization.property("domain_create", core.serialization.boolean().optional()),
    domainUpdate: core.serialization.property("domain_update", core.serialization.boolean().optional()),
    domainDelete: core.serialization.property("domain_delete", core.serialization.boolean().optional()),
    listEntryRead: core.serialization.property("list_entry_read", core.serialization.boolean().optional()),
    listEntryCreate: core.serialization.property("list_entry_create", core.serialization.boolean().optional()),
    listEntryDelete: core.serialization.property("list_entry_delete", core.serialization.boolean().optional()),
    metricsRead: core.serialization.property("metrics_read", core.serialization.boolean().optional()),
    apiKeyRead: core.serialization.property("api_key_read", core.serialization.boolean().optional()),
    apiKeyCreate: core.serialization.property("api_key_create", core.serialization.boolean().optional()),
    apiKeyDelete: core.serialization.property("api_key_delete", core.serialization.boolean().optional()),
    podRead: core.serialization.property("pod_read", core.serialization.boolean().optional()),
    podCreate: core.serialization.property("pod_create", core.serialization.boolean().optional()),
    podDelete: core.serialization.property("pod_delete", core.serialization.boolean().optional()),
});
