// agentmail/dist/cjs/Client.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentMailClient = void 0;
const Client_js_1 = require("./api/resources/agent/client/Client.js");
const Client_js_2 = require("./api/resources/apiKeys/client/Client.js");
const Client_js_3 = require("./api/resources/auth/client/Client.js");
const Client_js_4 = require("./api/resources/domains/client/Client.js");
const Client_js_5 = require("./api/resources/drafts/client/Client.js");
const Client_js_6 = require("./api/resources/inboxes/client/Client.js");
const Client_js_7 = require("./api/resources/lists/client/Client.js");
const Client_js_8 = require("./api/resources/metrics/client/Client.js");
const Client_js_9 = require("./api/resources/organizations/client/Client.js");
const Client_js_10 = require("./api/resources/pods/client/Client.js");
const Client_js_11 = require("./api/resources/threads/client/Client.js");
const Client_js_12 = require("./api/resources/webhooks/client/Client.js");
const Client_js_13 = require("./api/resources/websockets/client/Client.js");
const BaseClient_js_1 = require("./BaseClient.js");
const core = __importStar(require("./core/index.js"));
const environments = __importStar(require("./environments.js"));
class AgentMailClient {
    constructor(options = {}) {
        this._options = (0, BaseClient_js_1.normalizeClientOptionsWithAuth)(options);
    }
    get inboxes() {
        var _a;
        return ((_a = this._inboxes) !== null && _a !== void 0 ? _a : (this._inboxes = new Client_js_6.InboxesClient(this._options)));
    }
    get pods() {
        var _a;
        return ((_a = this._pods) !== null && _a !== void 0 ? _a : (this._pods = new Client_js_10.PodsClient(this._options)));
    }
    get webhooks() {
        var _a;
        return ((_a = this._webhooks) !== null && _a !== void 0 ? _a : (this._webhooks = new Client_js_12.WebhooksClient(this._options)));
    }
    get agent() {
        var _a;
        return ((_a = this._agent) !== null && _a !== void 0 ? _a : (this._agent = new Client_js_1.AgentClient(this._options)));
    }
    get apiKeys() {
        var _a;
        return ((_a = this._apiKeys) !== null && _a !== void 0 ? _a : (this._apiKeys = new Client_js_2.ApiKeysClient(this._options)));
    }
    get auth() {
        var _a;
        return ((_a = this._auth) !== null && _a !== void 0 ? _a : (this._auth = new Client_js_3.AuthClient(this._options)));
    }
    get domains() {
        var _a;
        return ((_a = this._domains) !== null && _a !== void 0 ? _a : (this._domains = new Client_js_4.DomainsClient(this._options)));
    }
    get drafts() {
        var _a;
        return ((_a = this._drafts) !== null && _a !== void 0 ? _a : (this._drafts = new Client_js_5.DraftsClient(this._options)));
    }
    get lists() {
        var _a;
        return ((_a = this._lists) !== null && _a !== void 0 ? _a : (this._lists = new Client_js_7.ListsClient(this._options)));
    }
    get metrics() {
        var _a;
        return ((_a = this._metrics) !== null && _a !== void 0 ? _a : (this._metrics = new Client_js_8.MetricsClient(this._options)));
    }
    get organizations() {
        var _a;
        return ((_a = this._organizations) !== null && _a !== void 0 ? _a : (this._organizations = new Client_js_9.OrganizationsClient(this._options)));
    }
    get threads() {
        var _a;
        return ((_a = this._threads) !== null && _a !== void 0 ? _a : (this._threads = new Client_js_11.ThreadsClient(this._options)));
    }
    get websockets() {
        var _a;
        return ((_a = this._websockets) !== null && _a !== void 0 ? _a : (this._websockets = new Client_js_13.WebsocketsClient(this._options)));
    }
    /**
     * Make a passthrough request using the SDK's configured auth, retry, logging, etc.
     * This is useful for making requests to endpoints not yet supported in the SDK.
     * The input can be a URL string, URL object, or Request object. Relative paths are resolved against the configured base URL.
     *
     * @param {Request | string | URL} input - The URL, path, or Request object.
     * @param {RequestInit} init - Standard fetch RequestInit options.
     * @param {core.PassthroughRequest.RequestOptions} requestOptions - Per-request overrides (timeout, retries, headers, abort signal).
     * @returns {Promise<Response>} A standard Response object.
     */
    fetch(input, init, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            return core.makePassthroughRequest(input, init, {
                baseUrl: (_a = this._options.baseUrl) !== null && _a !== void 0 ? _a : (() => __awaiter(this, void 0, void 0, function* () {
                    var _a;
                    const env = yield core.Supplier.get(this._options.environment);
                    return typeof env === "string"
                        ? env
                        : ((_a = env === null || env === void 0 ? void 0 : env.http) !== null && _a !== void 0 ? _a : environments.AgentMailEnvironment.Prod.http);
                })),
                headers: this._options.headers,
                timeoutInSeconds: this._options.timeoutInSeconds,
                maxRetries: this._options.maxRetries,
                fetch: this._options.fetch,
                logging: this._options.logging,
                getAuthHeaders: () => __awaiter(this, void 0, void 0, function* () { return (yield this._options.authProvider.getAuthRequest()).headers; }),
            }, requestOptions);
        });
    }
}
exports.AgentMailClient = AgentMailClient;
